Point at source of trait bound obligations in more places

estebank opened the pull request:
Be more thorough in using `ItemObligation` and `BindingObligation` when
evaluating obligations so that we can point at trait bounds that
introduced unfulfilled obligations. We no longer incorrectly point at
unrelated trait bounds (`substs-ppaux.verbose.stderr`).

In particular, we now point at trait bounds on method calls.

We no longer point at "obvious" obligation sources (we no longer have a
note pointing at `Trait` saying "required by a bound in `Trait`", like
in `associated-types-no-suitable-supertrait*`).

We no longer point at associated items (`ImplObligation`), as they didn't
add any user actionable information, they just added noise.

Address part of #89418.

estebank committed:
Point at source of trait bound obligations in more places

Be more thorough in using `ItemObligation` and `BindingObligation` when
evaluating obligations so that we can point at trait bounds that
introduced unfulfilled obligations. We no longer incorrectly point at
unrelated trait bounds (`substs-ppaux.verbose.stderr`).

In particular, we now point at trait bounds on method calls.

We no longer point at "obvious" obligation sources (we no longer have a
note pointing at `Trait` saying "required by a bound in `Trait`", like
in `associated-types-no-suitable-supertrait*`).

Address part of #89418.

rust-highfive commented:
r? @nagisa

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3808995925) [(plain)](https://github.com/rust-lang/rust/commit/8afc5f9a527ef633f522e3690d860e3a72fd390a/checks/3808995925/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [incremental] incremental/hashes/trait_defs.rs stdout ----

error in revision `cfail5`: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/incremental/hashes/trait_defs.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "cfail5" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/trait_defs.inc" "-Z" "incremental-verify-ich" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "query-dep-graph" "-Zincremental-relative-spans" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error: `predicates_of(TraitVisibility)` should be dirty but is not
   |
   |
LL | pub trait TraitVisibility { }


error: `predicates_of(TraitUnsafety)` should be dirty but is not
   |
   |
LL | unsafe trait TraitUnsafety { }


error: `predicates_of(TraitAddMethod)` should be dirty but is not
   |
   |
LL | / pub trait TraitAddMethod {
LL | |     fn method();
LL | | }

error: aborting due to 3 previous errors


---
test result: FAILED. 145 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 11.16s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/incremental" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "incremental" "--mode" "incremental" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:11:30
```

</details>

estebank committed:
Change `trait_defs.rs` incremental hash test

`predicates_of` no longer changes when changing a trait's front matter
because we no longer include the trait's span in the identity trait
obligation.

estebank committed:
Point at bounds when comparing impl items to trait

cjgillot commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -2138,10 +2138,19 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+        let span = match generics {
+            // This is to get around the trait identity obligation, that has a `DUMMY_SP` as signal
+            // for other diagnostics, so we need to recover it here.
+            Some((_, _, node)) if span.is_dummy() => node,
```
Should this test be performed around the `hir.span(hir_id)` call?

cjgillot commented:
> @cjgillot, I see that you touched this file. Do you know if there's _any_ reliance on `predicates_of` changing when `trait` modifiers like `pub` and `unsafe` are changed? Currently they do because the change of the span, but with this PR's change, that's no longer true. Would it be reasonable to change the test to no longer look for that?

I don't know enough the trait solving algorithm to comment on the need for the value to change when pub and unsafe change. OTOH, I would consider that any incremental code that relies on span-related invalidation to work correctly would be a red flag. So IMHO, it's fine to update the test.

If you want confirmation, you'd need to ask to professional trait solvers like @jackh726.

jackh726 commented:
> If you want confirmation, you'd need to ask to professional trait solvers like @jackh726.
😂❤you give me too much credit

> Do you know if there's _any_ reliance on `predicates_of` changing when `trait` modifiers like `pub` and `unsafe` are changed? 
I would be extremely surprised if this were the case. I imagine for everything but diagnostics, we throw away the span *anyways*.

nagisa reviewed (commented):
Looked through the test changes for now, there seem to be some unfortunate regressions at least from my perspective.

estebank committed:
Do not mention associated items when they introduce an obligation

estebank committed:
Align multiline messages to their label (add left margin)

estebank committed:
Point at `impl` blocks when they introduce unmet obligations

Group obligations by `impl` block that introduced them.

estebank committed:
Suggest constraining `fn` type params when appropriate

estebank committed:
Move tests for missing trait bounds to their own directory

bors commented:
:umbrella: The latest upstream changes (presumably #89903) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nagisa reviewed (approved):
This seems very nice. r=me with one remaining nit.

estebank commented on compiler/rustc_typeck/src/check/compare_method.rs:
```diff
@@ -227,12 +223,22 @@ fn compare_predicate_entailment<'tcx>(
 
             inh.register_predicates(obligations);
-            inh.register_predicate(traits::Obligation::new(cause.clone(), param_env, predicate));
+            let cause = ObligationCause::new(
```
Done.

estebank committed:
review comments

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4246055332?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/2c173afb7e72d305aeb16d7db9a863eaea26c51a/checks/4246055332/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Finished release [optimized] target(s) in 9.65s
tidy check
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
tidy error: following path contains more than 1102 entries, you should move the test to some relevant subdirectory (current: 1103): /checkout/src/test/ui
* highest error code: E0786
Found 502 error codes
Found 0 error codes with no tests
Done!
Done!
* 358 features
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:12
```

</details>

estebank committed:
Move tests from ui directory

estebank commented:
@bors r=nagisa

bors commented:
:pushpin: Commit a8dcc87041db9ab0eb4865a42baa8eb9b0a4a7fc has been approved by `nagisa`

<!-- @bors r=nagisa a8dcc87041db9ab0eb4865a42baa8eb9b0a4a7fc -->
<!-- homu: {"type":"Approved","sha":"a8dcc87041db9ab0eb4865a42baa8eb9b0a4a7fc","approver":"nagisa"} -->

Mark-Simulacrum commented on compiler/rustc_errors/src/emitter.rs:
```diff
@@ -1266,22 +1266,37 @@ impl EmitterWriter {
                 buffer.append(0, "[", Style::Level(*level));
                 buffer.append(0, &code, Style::Level(*level));
                 buffer.append(0, "]", Style::Level(*level));
+                label_width += 2 + code.len();
```
Can we say something like `": ".len()` presuming I've guessed correctly as to the meaning of this 2? (Or just a comment on at least one of these).

estebank replied:
This `2` belongs to the two `[` `]` surrounding the `code`. There's another `2` below for the `": "`. Kept these magic numbers very close to their causes to try and make them obvious, but I guess they aren't?

bors commented:
:umbrella: The latest upstream changes (presumably #91080) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank commented:
@bors r=nagisa

bors commented:
:pushpin: Commit d5e982d6894ea34a19a0e31eb484976a07085030 has been approved by `nagisa`

<!-- @bors r=nagisa d5e982d6894ea34a19a0e31eb484976a07085030 -->
<!-- homu: {"type":"Approved","sha":"d5e982d6894ea34a19a0e31eb484976a07085030","approver":"nagisa"} -->

estebank committed:
bless NLL test

estebank commented:
@bors r=nagisa

bors commented:
:pushpin: Commit 2a2621deef74e6e9687f96ea502c7f82b1ffb432 has been approved by `nagisa`

<!-- @bors r=nagisa 2a2621deef74e6e9687f96ea502c7f82b1ffb432 -->
<!-- homu: {"type":"Approved","sha":"2a2621deef74e6e9687f96ea502c7f82b1ffb432","approver":"nagisa"} -->

estebank commented on src/test/ui/associated-types/substs-ppaux.verbose.stderr:
```diff
@@ -82,11 +82,6 @@ note: required because of the requirements on the impl of `Foo<'_#0r, '_#1r, u8>
    |
 LL | impl<'a,'b,T,S> Foo<'a, 'b, S> for T {}
    |                 ^^^^^^^^^^^^^^     ^
-note: required by a bound in `Foo::bar`
-  --> $DIR/substs-ppaux.rs:7:30
-   |
-LL |     fn bar<'a, T>() where T: 'a {}
-   |                              ^^ required by this bound in `Foo::bar`
```
This looks like a regression :(

bors commented:
:hourglass: Testing commit 2a2621deef74e6e9687f96ea502c7f82b1ffb432 with merge b8e5ab20ed7a7677a998a163ccf7853764b195e6...
<!-- homu: {"type":"BuildStarted","head_sha":"2a2621deef74e6e9687f96ea502c7f82b1ffb432","merge_sha":"b8e5ab20ed7a7677a998a163ccf7853764b195e6"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4277830246?check_suite_focus=true)
Approved by: nagisa
Pushing b8e5ab20ed7a7677a998a163ccf7853764b195e6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4277830246?check_suite_focus=true"},"merge_sha":"b8e5ab20ed7a7677a998a163ccf7853764b195e6"} -->

rust-timer commented:
Finished benchmarking commit (b8e5ab20ed7a7677a998a163ccf7853764b195e6): [comparison url](https://perf.rust-lang.org/compare.html?start=02913c078849f940371eb9930754f2b0f1bc9fad&end=b8e5ab20ed7a7677a998a163ccf7853764b195e6).

**Summary**: This change led to large relevant mixed results 🤷 in compiler performance.
- Large improvement in instruction counts (up to -5.0% on `incr-unchanged` builds of `inflate`)
- Small regression in instruction counts (up to 1.1% on `incr-unchanged` builds of `wg-grammar`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression

pnkfelix commented:
Visiting for weekly performance triage.

Many other PR's over the past week have observed a similar delta to `incr-unchanged` builds of `inflate`, so I am dismissing *that* change as spurious. (The current hypothesis is that it is due to https://github.com/rust-lang/rustc-perf/issues/1105.)

Other than that, there were a broad set of small regressions. Putting aside the ones tagged with `?` ("noisy"), there are 19 benchmarks that regressed by 0.10% to 0.42%. This seems like an acceptable cost.

@rustbot label: +perf-regression-triaged
